pkgs/build-support/dotnet/build-dotnet-module/default.nix

```diff
@@ -134,45 +150,87 @@ in stdenvNoCC.mkDerivation (args // {
-        dotnet restore "$project" \
-          ${lib.optionalString (!enableParallelBuilding) "--disable-parallel"} \
-          -p:ContinuousIntegrationBuild=true \
-          -p:Deterministic=true \
-          --packages "$HOME/nuget_pkgs" \
-          ${lib.optionalString (dotnetRestoreFlags != []) (builtins.toString dotnetRestoreFlags)} \
-          ${lib.optionalString (dotnetFlags != []) (builtins.toString dotnetFlags)}
-      done
-
-      echo "${lib.concatStringsSep "\n" exclusions}" > "$HOME/package_exclusions"
-
-      echo "Writing lockfile..."
-      nuget-to-nix "$HOME/nuget_pkgs" "$HOME/package_exclusions" > "$deps_file"
-      echo "Succesfully wrote lockfile to: $deps_file"
-    '';
-  } // args.passthru or {};
+    fetch-deps =
+      let
+        # Because this list is rather long its put in its own store path to maintain readability of the generated script
+        exclusions = writeText "nuget-package-exclusions" (lib.concatStringsSep "\n" (dotnet-sdk.passthru.packages { fetchNuGet = attrs: attrs.pname; }));
+
+        # Derivations may set flags such as `--runtime <rid>` based on the host platform to avoid restoring/building nuget dependencies they dont have or dont need.
+        # This introduces an issue; In this script we loop over all platforms from `meta` and add the RID flag for it, as to fetch all required dependencies.
+        # The script would inherit the RID flag from the derivation based on the platform building the script, and set the flag for any iteration we do over the RIDs.
+        # That causes conflicts. To circumvent it we remove all occurances of the flag.
+        flags =
+          let
+            hasRid = flag: builtins.any (v: v) (map (rid: lib.hasInfix rid flag) (lib.attrValues dotnet-sdk.runtimeIdentifierMap));
+          in
+          builtins.filter (flag: !(hasRid flag)) (dotnetFlags ++ dotnetRestoreFlags);
+
+        runtimeIds = map (system: dotnet-sdk.systemToDotnetRid system) (args.meta.platforms or dotnet-sdk.meta.platforms);
+      in
+      writeShellScript "fetch-${pname}-deps" ''
+        set -euo pipefail
+
+        export PATH="${lib.makeBinPath [ coreutils dotnet-sdk nuget-to-nix ]}"
+
+        case "''${1-}" in
+            --help|-h)
+                echo "usage: $0 <output path> [--help]"
+                echo "    <output path>  The path to write the lockfile to"
+                echo "    --help         Show this help message"
+                exit
+                ;;
+        esac
+
+        deps_file="$(realpath "''${1:-$(mktemp -t "XXXXXX-${pname}-deps.nix")}")"
+        export HOME=$(mktemp -td "XXXXXX-${pname}-home")
+        mkdir -p "$HOME/nuget_pkgs"
+
+        store_src="${srcOnly args}"
+        src="$(mktemp -td "XXXXXX-${pname}-src")"
+        cp -rT "$store_src" "$src"
+        chmod -R +w "$src"
+        trap "rm -rf $src $HOME" EXIT
```

Comment:
Not sure if we should clean up when erroring or the script was aborted.

Reply:
I'd say its preferable to keep that as the default behavior. In my experience a lot of errors are trivially fixable from the derivation, inspecting the state of the temporary files is not very useful a lot of the time. Having to manually delete the files in upon any error/abort seems like a bit of a pain.

Maybe we should add an `--keep-source` flag or something that would disable deleting them? That could also print the paths in the trap.

Reply:
If we want to clean up the files all the time (which I think we should) I think we need to add INT and TERM.